include/execution.hpp

```diff
@@ -1873,6 +1860,251 @@ namespace std::execution {
+                  // All child operations completed successfully:
+                  std::apply(
+                    [this](auto&... opt_values) -> void {
+                      std::apply(
+                        [this](auto&... all_values) -> void {
+                          execution::set_value(
+                              (Receiver&&) __recvr_, std::move(all_values)...);
+                        },
+                        std::tuple_cat(
+                          std::apply(
+                            [this](auto&... vals) { return std::tie(vals...); },
+                            *opt_values
+                          )...
+                        )
+                      );
+                    },
+                    __values_
+                  );
+                  break;
+                case __error:
+                  std::visit([this](auto& error) noexcept {
+                    execution::set_error((Receiver&&) __recvr_, std::move(error));
+                  }, __errors_);
+                  break;
+                case __done:
+                  execution::set_done((Receiver&&) __recvr_);
+                  break;
+                default:
+                  ;
+                }
+              }
+
+              struct __completion_fn {
+                void operator()() const noexcept {
+                  __op_->complete();
+                }
+                __operation* __op_;
+              };
+
+              __operation(WhenAll&& when_all, Receiver recvr)
+                : __child_states_{connect_children(this, (WhenAll&&) when_all, Indices{})}
+                , __recvr_((Receiver&&) recvr)
+                , __barrier_(sizeof...(SenderIds), __completion_fn{this})
+              {}
+
+              friend void tag_invoke(start_t, __operation& self) noexcept {
+                // register stop callback:
+                self.__on_stop_.emplace(
+                    get_stop_token(self.__recvr_),
+                    __on_stop_requested{self.__stop_source_});
+                if (self.__stop_source_.stop_requested()) {
+                  // Stop has already been requested. Don't bother starting
+                  // the child operations.
+                  execution::set_done((Receiver&&) self.__recvr_);
+                } else {
+                  apply([](auto&&... child_ops) noexcept -> void {
+                    (execution::start(child_ops), ...);
+                  }, self.__child_states_);
+                }
+              }
+
+              child_optstates_tuple_t __child_states_;
+              Receiver __recvr_;
+              barrier<__completion_fn> __barrier_;
+              // Could be non-atomic here and atomic_ref everywhere except __completion_fn
+              atomic<__state_t> __state_{__started};
+              error_types<variant> __errors_{};
+              tuple<typename sender_traits<__t<SenderIds>>::template
+                  value_types<tuple, optional>...> __values_{};
```

Comment:
I could save some space here by putting the values and errors into one variant, but I would need to figure out how to "lock" the variant while writing values into it so that if another thread errors, it doesn't try to concurrently access it.